assets/javascripts/kitten/components/images/image-cropper.js

```diff
@@ -0,0 +1,304 @@
+    this.handleUploaderReset = this.handleUploaderReset.bind(this)
+
+    this.handleSliderChange = this.handleSliderChange.bind(this)
+    this.handleSliderAction = this.handleSliderAction.bind(this)
+
+    this.handleReady = this.handleReady.bind(this)
+    this.handleCrop = this.handleCrop.bind(this)
+
+    this.renderError = this.renderError.bind(this)
+    this.setCropperHeight = this.setCropperHeight.bind(this)
+  }
+
+  initialState() {
+    return {
+      imageSrc: null,
+      imageCropSrc: null,
+      fileName: null,
+      touched: false,
+      sliderValue: 0,
+      sliderMin: this.props.sliderMin,
+      sliderMax: this.props.sliderMax,
+    }
+  }
+
+  componentDidMount() {
+    this.setCropperHeight()
+
+    window.addEventListener('resize', this.setCropperHeight)
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.setCropperHeight)
+  }
+
+  handleUploaderSuccess(data) {
+    this.setState({
+      imageSrc: data.preview,
+      imageCropSrc: null,
+      fileName: data.name,
+      sliderValue: 0,
+    }, this.setCropperHeight)
+  }
+
+  handleUploaderError(hasError) {
+    const resetState = hasError ? this.initialState() : {}
+
+    this.setState({
+      hasErrorOnUploader: hasError,
+      ...resetState,
+    })
+  }
+
+  handleUploaderReset() {
+    this.setState(this.initialState())
+  }
+
+  handleSliderChange(value) {
+    this.setState({
+      sliderValue: value,
+    })
+
+    this.refs.cropper.zoomTo(value / 100)
+  }
+
+  handleSliderAction() {
+    this.setState({ touched: true })
+  }
+
+  // Calculate the right range for the zoom slider.
+  handleReady() {
+    const imageData = this.refs.cropper.getImageData()
+    const naturalWidth = imageData.naturalWidth
+    const width = imageData.width
+    const ratio = width / naturalWidth * 100
+    const min = this.props.sliderMin + ratio
+    const max = this.props.sliderMax + ratio
+
+    this.setState({
+      sliderMin: min,
+      sliderMax: max,
+      sliderValue: min,
+    })
+  }
+
+  handleCrop() {
+    if (!this.state.imageSrc) return
+
+    const croppedCanvas = this.refs.cropper.getCroppedCanvas()
+
+    if (croppedCanvas) {
+      const imageCropSrc = croppedCanvas.toDataURL()
+
+      this.setState({
+        imageCropSrc: imageCropSrc,
+      })
+
+      this.props.onChange({
+        value: imageCropSrc,
+        name: this.state.fileName,
+      })
+    }
+  }
+
+  setCropperHeight() {
+    if (this.cropperContainer) {
+      const width = domElementHelper.getComputedWidth(this.cropperContainer)
+      const height = width / this.props.aspectRatio
+
+      this.setState({
+        cropperWidth: width,
+        cropperHeight: height,
+      })
+    }
+  }
+
+  renderCropper() {
+    const styles = {
+      width: this.state.cropperWidth,
+      height: this.state.cropperHeight,
+    }
+
+    return (
+      <Marger key="cropper" top="2" bottom="2">
+        <div ref={ node => { this.cropperContainer = node } }>
```

Comment:
Je viens de voir que dans la doc de React, on a les deux : 
- avec le `_` : https://zhenyong.github.io/react/docs/more-about-refs.html,
- sans : https://facebook.github.io/react/docs/refs-and-the-dom.html.